### PR TITLE
classlib: LinuxPlatform: findTerminalEmulator

### LINKED ISSUE #4893: String.runInTerminal compatibility issue across Linux distros
* SuperCollider version: 3.10.4
* Operating system: Linux Mint 19.3 Tricia
* Other details (Qt version, audio driver, etc.): 

## Steps to reproduce

```supercollider
"ls".runInTerminal; // this command has no effect on Linux Mint
```

## Expected vs. actual behavior

The implementation of String.runInTerminal in exString_linux.sc looks like this:
```

+ String {
	runInTerminal {|shell="sh"|
		("xterm -hold -T 'SuperCollider runInTerminal' -e" + shell + "-c" + this.shellQuote).unixCmd;
	}

	openOS {
		("xdg-open " ++ this.escapeChar($ )).systemCmd
	}
```
The issue is that Linux Mint (and other distros) uses gnome-terminal by default and not xterm. In fact Mint doesn't come packaged with xterm at all. runInTerminal fails silently.

It would be great if this command used the system default terminal out of the box - maybe via a switch statement that checks against some common distros? Or at least should check if xterm exists on the system, or give the option to pass the terminal application choice into the method.

### PR DESCRIPTION
**Pull request by elgiano**
## Purpose and Motivation

Fixes #4893
## Types of changes

- Documentation
- New feature

## To-do list

- [x] Code is tested
- [x] All tests are passing
- [x] Updated documentation
- [x] This PR is ready for review

### CODE REVIEW

File: SCClassLibrary/Common/Collections/linux/extLinuxPlatform_findTerminalEmulator.sc
Comment (dyfer) [inline SCClassLibrary/Common/Collections/linux/extLinuxPlatform_findTerminalEmulator.sc:4]: I have 3 comments here:

- I'd prefer consistent symbol notation - in this case ```'symbol'``` and not ```\symbol```
- rethink the order of commands prioritizing more common options
  - for me that's rather theoretical since I'm not using Linux regularly, but my guess is that we should do something like:
    1. universal terminal command first (*)
    2. gnome/kde/other desktop environments
    3. xterm and other popular emulators
- re: (*): I understand that there's no universal terminal command, that would launch terminal of choice on the user machine, but I read that there is `x-terminal-emulator`. IIUC this would be the closest option to "open default terminal", even though it's only for debian, AFAIU.

File: SCClassLibrary/Common/Collections/linux/extString_linux.sc
Comment (elgiano) [reply SCClassLibrary/Common/Collections/linux/extString_linux.sc:7]: Agreed, we can just mention in the docs that String:runInTerminal on Linux calls findTerminalEmulator first, if it hadn't been  called already

File: SCClassLibrary/Common/Collections/linux/extLinuxPlatform_findTerminalEmulator.sc
Comment (elgiano) [reply SCClassLibrary/Common/Collections/linux/extLinuxPlatform_findTerminalEmulator.sc:4]: - Consistent symbol notation: agreed!
- Reorder: right now there's no defined order, am I right? They are dictionary keys, so their order is not reliable. I'll think about this, as there should also be a way for users to set their preferred emulator as first priority

Comment (dyfer) [reply SCClassLibrary/Common/Collections/linux/extLinuxPlatform_findTerminalEmulator.sc:4]: > * Reorder: right now there's no defined order, am I right? They are dictionary keys, so their order is not reliable. I'll think about this, as there should also be a way for users to set their preferred emulator as first priority

Ah yes, I've read it wrong.
When I was thinking about the implementation, I was considering having `supportedTerminalEmulatorsCmds` as an array of commands, and then use `string.spli($ ).first` or something like that to get the command name itself when iterating to test their validity. That way the commands can be set in preferred order. What do you think about that?

File: SCClassLibrary/Common/Collections/linux/extLinuxPlatform_findTerminalEmulator.sc
Comment (brianlheim) [inline SCClassLibrary/Common/Collections/linux/extLinuxPlatform_findTerminalEmulator.sc:36]: please use tabs in all SC code :)

File: SCClassLibrary/Common/Collections/linux/extString_linux.sc
Comment (brianlheim) [inline SCClassLibrary/Common/Collections/linux/extString_linux.sc:8]: these inner blocks are twice-indented

File: SCClassLibrary/Common/Collections/linux/extLinuxPlatform_findTerminalEmulator.sc
Comment (elgiano) [reply SCClassLibrary/Common/Collections/linux/extLinuxPlatform_findTerminalEmulator.sc]: Makes sense! Now, three problems:
- x-terminal-emulator: we can't predict the right arguments' syntax for it, since it's a symlink and we don't know what exact program it points to. I think it's better to avoid it.
- terminals without --hold option (xfce, gnome, mate, and terminator): I don't know what's best to do, as I feel this is an important feature of runInTerminal, and appending `;read` to user commands doesn't feel good to me. Worse, the ones without --hold are probably among the most used on debian-based distros. I gave them lower priority for now.
- some terminals seem to need extra quotes around the shell command ( xfce, mate, termite, terminator )... which apparently sends us to an escaping hell... there must be some other way, but I haven't found it yet, just leaving this as a note

File: SCClassLibrary/Common/Collections/linux/extLinuxPlatform_findTerminalEmulator.sc
Comment (dyfer) [inline SCClassLibrary/Common/Collections/linux/extLinuxPlatform_findTerminalEmulator.sc:13]: Isn't gnome-terminal's switch for command `-e` instead of `-c`? Just reading the man page, haven't tested

File: SCClassLibrary/Common/Collections/linux/extLinuxPlatform_findTerminalEmulator.sc
Comment (dyfer) [reply SCClassLibrary/Common/Collections/linux/extLinuxPlatform_findTerminalEmulator.sc]: > * x-terminal-emulator: we can't predict the right arguments' syntax for it, since it's a symlink and we don't know what exact program it points to. I think it's better to avoid it.

Got it... I didn't realize this is just a symlink. Then yes, better to avoid it.

> * terminals without --hold option (...)

What if we went the "temporary script" route, like on macOS? I _think_ this would help with escaping.

> * some terminals seem to need extra quotes around the shell command  (...)

Based on [this](https://stackoverflow.com/questions/3512055/avoid-gnome-terminal-close-after-script-execution) we could consider dropping to `bash` after the command... what do you think?

**Comment by brianlheim** [comment]
> please use tabs in all SC code :)

@elgiano your last commit switched everything to spaces, FYI.

**Comment by elgiano** [comment]
Ops I definitely have to double check my scide settings!

File: SCClassLibrary/Common/Collections/linux/extLinuxPlatform_findTerminalEmulator.sc
Comment (elgiano) [reply SCClassLibrary/Common/Collections/linux/extLinuxPlatform_findTerminalEmulator.sc:13]: From gnome-terminal's output
# Option “-e” is deprecated and might be removed in a later version of gnome-terminal.
# Use “-- ” to terminate the options and put the command line to execute after it.

Comment (dyfer) [reply SCClassLibrary/Common/Collections/linux/extLinuxPlatform_findTerminalEmulator.sc:13]: Oh go it, sorry for the noise!

File: SCClassLibrary/Common/Collections/linux/extLinuxPlatform_findTerminalEmulator.sc
Comment (elgiano) [reply SCClassLibrary/Common/Collections/linux/extLinuxPlatform_findTerminalEmulator.sc]: I'm giving it a try with temp script and drop to shell.
- temp script: I think it's nice to help with escaping, and it makes terminal commands easier. Plus it's the same thing we do on macos.
- drop to shell vs. `read`:  dropping to a shell feels a bit too heavy handed. I'm giving it a try, but I think I would prefer to append something like `read -p "Press ENTER to close"`
- I would then remove all --hold flags from other terminal commands, to unify the behavior

**Comment by elgiano** [comment]
Thanks @dyfer and @brianlheim for your inputs, checks and comments.
I consider myself satisfied with this implementation (temp script + append `read`). If there aren't already any other issues, I have two doubts left before moving on to writing tests and documentation:
- LinuxPlatform.runInTerminalCmd: does this variable name make sense? I think it's clear and points clearly to .runInTerminal, but maybe it's a bit too verbose? Any thoughts/alternatives?
- LinuxPlatform.findTerminalEmulator: should we move this from its own extension file to the main LinuxPlatform class file?

**Comment by dyfer** [comment]
Thanks @elgiano 

I think in general this looks pretty good.

I really would like to revisit the topic of sync vs async behavior for finding the command. Did you have a chance to test or compare performance? Does forking this process is actually less disruptive for e.g. timing of other routines than doing it synchronously?

If we did this synchronously, in `String runInTerminal {` we could do:

```supercollider
...
LinuxPlatform.runInTerminalCmd ?? {LinuxPlatform.findTerminalEmulator};
LinuxPlatform.runInTerminalCmd !? {
// run in terminal code
}
...
```
Optionally `LinuxPlatform.findTerminalEmulator` could return `true`/`false` if the terminal is found.
Or we could even have `.findTerminalEmulator` run automatically (in LinuxPlatform method) when `runInTerminalCmd` is requested and is `nil`.

Again, what you have is functional, but I wonder if we can avoid this duality
`LinuxPlatform.runInTerminalCmd.format(title, fpath).unixCmd`
vs
`LinuxPlatform.findTerminalEmulator { |termCmd| termCmd...unixCmd }`

> 
> * LinuxPlatform.runInTerminalCmd: does this variable name make sense? I think it's clear and points clearly to .runInTerminal, but maybe it's a bit too verbose? Any thoughts/alternatives?

I think it's good. Maybe an alternative would be `LinuxPlatform.terminalCmd` but that might be misleading because it's a special string with argument

> * LinuxPlatform.findTerminalEmulator: should we move this from its own extension file to the main LinuxPlatform class file?

I'm split on this one. Even if it's a separate file, maybe it should go to `Platform/linux`? But then ext_LinuxPlatform would be next to LinuxPlatform, which probably doesn't make sense. I'm leaning towards including it in the regular LinuxPlatform. In comparison I see that OSXPlatform has at least one "random" method in it (`-writeClientCSS`). I wonder if @brianlheim has better ideas about this.

Also, looking at other SC class library code I don't see any ...Platforms (LinuxPlatform, OSXPlatform, WindowsPlatform) being explicitly called, it always seems to be just `Platform`. Maybe this is what should be used here as well when calling the methods (i.e. `Platform.runInTerminalCmd`)?

**Comment by elgiano** [comment]
> forking this process is actually less disruptive for e.g. timing of other routines than doing it synchronously?
Sorry, you had already mentioned it and I totally forgot. I assume it would be less disruptive to fork, and it makes sense to me since running the terminal will be async anyway (.unixCmd), but I can very well be wrong, so I'll do some checks!

> Or we could even have .findTerminalEmulator run automatically (in LinuxPlatform method) when runInTerminalCmd is requested and is nil.
Sure, isn't this already how it is?

> Again, what you have is functional, but I wonder if we can avoid this duality
Sure, now that you make me think again about it, we can already avoid this duality in the async version, I'm making a quick change for it now

> Platform.runInTerminalCmd
I would avoid it, since this is a Linux-only function, and Platform.runInTerminal sounds too cross-platform to me

thanks!

**Comment by elgiano** [comment]
@dyfer, and everyone interested
I pushed a temporary commit with both sync and async implementations. My tests showed that the sync versions, when only the last entry in the array is found, runs in an average of 0.04 seconds (that is by running .bench() for a thousand iterations, resetting LinuxPlatform.runInTerminalCmd to nil every time). However, when running a single .bench() manually, it can take up to 0.1 seconds, I don't know why.

I think this shouldn't anyway be a big concern, as one shouldn't look for a terminal emulator in the middle of some granular patterns (since you can do that at startup, before starting any other process, and this can be documented). But if the async code looks clean enough to you I would still prefer it, I've also cleaned it up to avoid code duplication :)

Let me know what you think! Thanks!

**Comment by dyfer** [comment]
Hi @elgiano 
Sorry for the delay.
I like how this is implemented (both sync and async versions), with the script writing similarly to macOS.
- have you tested all the terminals you've added? I've only tested xterm for now
- I have tested both sync and async version while running a sound-producing routine, and have not experienced noticeable timing disruption in either case (although maybe I was not pushing the system hard enough?)
  - I think both sync and async versions perform fine; my personal preference is to not fork when not necessary, but in this case forking probably doesn't hurt since the command is inherently asynchronous, and I can see the point of being potentially less disruptive. **I'd love someone else, particularly a linux user, to voice an opinion whether they prefer synchronous or asynchronous implementation of this**
- I think that `LinuxPlatform -findTerminalEmulator` should go to into the `Platform/linux/LinuxPlatform.sc` for two reasons: `runInTerminalCmd` is defined there, and you are adding to a (Linux)Platform, not a Collection. What do you think about moving this method there?

**Comment by elgiano** [comment]
Hi @dyfer! No problem and thanks for looking into this!
- I've tested all of them
- I agree it would be nice with at least a third person helping us decide between sync and async
- agree that we can move findTerminalEmulator to be next to runInTerminal

**Comment by brianlheim** [comment]
my 2c: do it synchronously. this is fully a convenience method, it can be a little slow and simple. i think even half a second is an acceptable delay for the first time running this.

on my system running i3 gnome-terminal was installed (possibly as a dependency of another package or because i forgot to install it during an experiment) and this method did not work well for me. there is also a cryptic message when xfce4-terminal is installed and found. however, all the other terminals on the list were compatible with my window manager. so, i would suggest putting the non-DE-specific terminals first, unless there's a good reason not to.

**Comment by dyfer** [comment]
Thanks for testing!

> i would suggest putting the non-DE-specific terminals first, unless there's a good reason not to.

I was suggesting to put IDE-specific terminals first, as these IMO would be more likely to hit a "default" terminal (and I assume that Gnome and KDE are still the most widely used DEs? ITK). But if that does't work well I won't press for that order.

**Comment by brianlheim** [comment]
> I was suggesting to put IDE-specific terminals first, as these IMO would be more likely to hit a "default" terminal (and I assume that Gnome and KDE are still the most widely used DEs? ITK). But if that does't work well I won't press for that order.

it's not a bad idea, but the error messages when you have a DE-specific terminal installed and it doesn't work are cryptic on my system, and if i didn't know this dynamic selection is happening i'd be very confused. the terminal chosen should work, and IIUC a non-DE-specific terminal emulator is more likely to work across DEs.

xfce4-terminal:

```
Failed to connect to session manager: Failed to connect to the session manager: SESSION_MANAGER environment variable not defined
```

gnome-terminal (after about 15 seconds of no activity):

```
# Error constructing proxy for org.gnome.Terminal:/org/gnome/Terminal/Factory0: Error calling StartServiceByName for org.gnome.Terminal: Timeout was reached
```

**Comment by elgiano** [comment]
I've chosen the sync version, moved relevant functions to LinuxPlatform.sc and sketched the docs.
I'll think about a couple of tests for this, but how does it work for platform specific testing?

File: HelpSource/Classes/LinuxPlatform.schelp
Comment (dyfer) [inline HelpSource/Classes/LinuxPlatform.schelp:14]: I think it would be better to use `code::runInTerminal::` instead `strong::runInTerminal::`

Comment (elgiano) [reply HelpSource/Classes/LinuxPlatform.schelp:14]: agreed, I'll change also `strong::nil::` then,

**Comment by dyfer** [comment]
I think this looks good now. 
I'm don't have any idea for a unit test for this, what do you think @elgiano?
@brianlheim what do you think about terminal choice now?

File: SCClassLibrary/Common/Collections/linux/extString_linux.sc
Comment (brianlheim) [inline SCClassLibrary/Common/Collections/linux/extString_linux.sc:10]: place these variables outside the `if` to avoid spurious inlining warnings

File: SCClassLibrary/Common/Collections/linux/extString_linux.sc
Comment (brianlheim) [inline SCClassLibrary/Common/Collections/linux/extString_linux.sc:7]: this is dead code; `error` throws so `^this` will never be reached. also, for future reference we usually don't put a space after `^`

File: SCClassLibrary/Common/Collections/linux/extString_linux.sc
Comment (brianlheim) [inline SCClassLibrary/Common/Collections/linux/extString_linux.sc:9]: why "string for terminal" when the function is called "runInTerminal"? obvious names are better

File: SCClassLibrary/Common/Collections/linux/extString_linux.sc
Comment (brianlheim) [inline SCClassLibrary/Common/Collections/linux/extString_linux.sc:21]: keep as runInTerminal -- programming is much simpler when the same name is used consistently across related objects. "SuperCollider in Terminal" could have come from anywhere, but "SuperCollider runInTerminal" (perhaps along with the first few characters of the executed string as well) makes the "cause" of the window unambiguously clear.

File: SCClassLibrary/Common/Collections/linux/extString_linux.sc
Comment (brianlheim) [reply SCClassLibrary/Common/Collections/linux/extString_linux.sc:7]: oh! you're right. in that case, i think this branch should actually throw, what do you think?

Comment (dyfer) [reply SCClassLibrary/Common/Collections/linux/extString_linux.sc:7]: >  i think this branch should actually throw, what do you think?

Theoretically yes, but on macOS we don't throw when there's a problem with temporary file creation, we similarly post the error without throwing.
While I don't have strong opinion, I'd indeed lean towards throwing an error and, in a separate PR, updating macOS implementation to also throw if there's a problem.

File: SCClassLibrary/Common/Collections/linux/extString_linux.sc
Comment (dyfer) [reply SCClassLibrary/Common/Collections/linux/extString_linux.sc:21]: I agree that `SuperCollider runInTerminal` is better. Incidentally, there doesn't seem to be a way to set the title in macOS terminal, or at least I can't figure it out (instead, it shows the filename of the script, which is `"SCStringForTerminal" ++ (this.hash.asString) ++ ".command";`).

File: SCClassLibrary/Common/Collections/linux/extString_linux.sc
Comment (dyfer) [reply SCClassLibrary/Common/Collections/linux/extString_linux.sc:9]: `SCStringForTerminal` is used in macOS implementation. I agree that `runInTerminal` would be better. (macOS can be changed later)

Comment (brianlheim) [reply SCClassLibrary/Common/Collections/linux/extString_linux.sc:9]: ah (re this and the comment above) -- i forgot there was a macOS impl. in that case, anything that's here to conform to the macOS impl can stay the way it is, i don't mind. sorry for the confusion!

File: SCClassLibrary/Common/Collections/linux/extString_linux.sc
Comment (elgiano) [reply SCClassLibrary/Common/Collections/linux/extString_linux.sc:7]: After reading the next comment, I left this as it was, calling .error and not throwing (like on mac). I liked the idea of throwing though, but I wouldn't like the same code to throw on one platform and not on another, so I think an eventual subsequent PR should change both implementations at the same time :)

File: SCClassLibrary/Common/Collections/linux/extString_linux.sc
Comment (elgiano) [reply SCClassLibrary/Common/Collections/linux/extString_linux.sc:21]: Changed to "SuperCollider runInTerminal"

**Comment by elgiano** [comment]
Sorry for the delay on this, I just addressed all review comments. Thanks again for looking at this! I feel we are close to finish it :)

**Comment by elgiano** [comment]
Sorry if my last comment was not really clear :)
I would like to have a final approval on this, then I will squash the commits, remove the WIP tag and feel be for merging

File: HelpSource/Classes/String.schelp
Comment (dyfer) [inline HelpSource/Classes/String.schelp:995]: Do we need this change? (it's minor, I know)

**Review by dyfer — CHANGES_REQUESTED** [review]
I tested on Ubuntu 20.04 and it picked gnome-terminal (great). However, the terminal doesn't stay open, seems like `read` has no effect on it for some reason... could you investigate that?
One alternative would be to drop to `bash` but that's not pretty either I guess...

edit: I tested xterm and it didn't work either - the command was found but the terminal didn't open. Seems like it still might need the `-hold` switch...?

File: HelpSource/Classes/String.schelp
Comment (elgiano) [reply HelpSource/Classes/String.schelp:995]: Definitely not needed, thanks for pointing that out!

**Comment by elgiano** [comment]
> I tested on Ubuntu 20.04 and it picked gnome-terminal (great). However, the terminal doesn't stay open, seems like `read` has no effect on it for some reason... could you investigate that?

Can you try that again? I found out I had a local commit that I hadn't push to origin... 
Anyway, I can't reproduce it on Arch w. GNOME 3, neither on wayland nor X11

**Comment by elgiano** [comment]
> I tested on Ubuntu 20.04 and it picked gnome-terminal (great). However, the terminal doesn't stay open, seems like `read` has no effect on it for some reason... could you investigate that?

Can you try that again? I found out I had a local commit that I hadn't push to origin... 
Anyway, I can't reproduce it on Arch w. GNOME 3, neither on wayland nor X11, before and after the missing commit

**Comment by dyfer** [comment]
thanks @elgiano 
I think what I'm seeing on my system is similar to this: https://stackoverflow.com/questions/15744421/read-command-doesnt-wait-for-input
Seems like we can't rely on `read` if we use `sh` (actually, the user can specify any shell)
My terminals were not opening from SC so I copied the path to system terminal and was trying to run. It did run but the terminal (`xterm`) wouldn't open, or rather, it wouldn't stay open. After I added `-hold` I saw this message:
`press ENTER to close this window./tmp/SCString....sh: 3: read: arg count`

So I guess the workaround in that StackOverflow response is one option. Also, is `gnome-terminal` the only one that doesn't have the command line switch to stay open?

Another idea: since we do specify the shell, maybe dropping to shell at the end is not such a bad idea?

**Comment by elgiano** [comment]
Thanks @dyfer 
Nice catch! I actually figured out that on my system `sh` is linked to `bash`. So, I've implemented the standard `read` as in the SO post you shared.
If I remember correctly, gnome-terminal is the only one in this list not implementing a hold switch.
I would prefer to avoid dropping to shell, but I agree that if things get more complicated it could be a solution

**Comment by dyfer** [comment]
cool, thanks! I'll test soon.
In the meantime, here are my thoughts:
- do we know if using standard `read` guaranteed to work in most shells?
- in my head I was entertaining the idea of using "hold" for terms that use it and implementing either `read` or dropping to shell for the ones that dont
  - pros:
    - cleaner implementation for most terminals
  - cons
    - inconsistent implementation between terminals
    - less pretty implementation

As for implementation, I'd think that 
- it could be a second class variable in LinuxPlatform, like `terminalCanHold` (haha?), 
- list of terminals would be an array of arrays with string as first arg and true/false as second
- `findTerminalEmulator` would set `terminalCanHold`
- `runInTerminal` would either add `read` or not, based on the value of `terminalCanHold`

What do you think? I'm not pushing either solution. I like using `-hold` where available, but I don't like inconsistent implementation.

**Comment by elgiano** [comment]
- this usage of read is POSIX standard, I'm sure it works on sh, bash, and zsh (tested). It should also work on csh and ksh. I feel like choosing the basic POSIX version is a good enough guarantee on our side.
- hold: I now think it's better to rely on POSIX read, instead of on program switches. It might even make it easier to add new terminals without worrying about how to hold them (or even if it's possible at all).
- everything else: honestly I wouldn't like to further articulate this implementation in anything more complicated than it already is :)

What do you think?

**Comment by dyfer** [comment]
This now works, thanks!

> * this usage of read is POSIX standard, I'm sure it works on sh, bash, and zsh (tested). It should also work on csh and ksh. I feel like choosing the basic POSIX version is a good enough guarantee on our side.
> * hold: I now think it's better to rely on POSIX read, instead of on program switches. It might even make it easier to add new terminals without worrying about how to hold them (or even if it's possible at all).

Sounds good!

> * everything else: honestly I wouldn't like to further articulate this implementation in anything more complicated than it already is :)

Yeah you're totally right! Sorry this takes so much back and forth. But it's going to be great! :)

I think this is almost ready, I have two little suggestions:

In help for `LinuxPlatform.findTerminalEmulator` I'd add the existing description:
```
returns:: command to invoke a terminal emulator, to be used as -runInTerminalCmd
```
or something like that. Not remembering how this is implemented, I had to look into class file to see whether this method returns the string, or internally sets `runInTerminalCmd`. Alternatively, what do you think about renaming it to `getTerminalEmulatorCmd` ? I don't feel strongly about it, but it might be more descriptive in a way what the method returns. Either way, I'd add the "returns" section to the help.

Also, I thought typically there's no space between `^` and the return value when returning from a method. However I don't see this being explicitly stated in our [code guide](https://github.com/supercollider/supercollider/wiki/Code-style-guidelines#return-statements) so feel free to leave it as is if you prefer.

**Comment by elgiano** [comment]
Agreed, all done!
Thanks for looking after this, and let me know if you find anything else that you feel like changing :+1:

**Review by dyfer — APPROVED** [review]
Thanks @elgiano 
IMO this is good to go now
I think it should be squashed on merge, unless you feel like rebasing to squash commits.
Thanks again for your work on this!

**Comment by elgiano** [comment]
Squashed, and ready! Thanks @dyfer and @brianlheim!

Now we just have to wait for the mac CI fix, and then we can finally merge this :)

**Comment by dyfer** [comment]
> Now we just have to wait for the mac CI fix, and then we can finally merge this :)

I don't personally see a problem merging this once Linux CIs pass; macOS CI failures are unrelated to this, especially that this is Linux-only functionality :)

**Comment by elgiano** [comment]
Agreed, feel free to merge then! Only mac CI is failing :)

**Review by brianlheim — APPROVED** [review]

**Pull request merged**
